crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -0,0 +1,599 @@
+    }
+
+    /// Returns a mutable reference to the data associated with the event that triggered the observer.
+    pub fn data_mut(&mut self) -> &mut E {
+        self.data
+    }
+
+    /// Returns a pointer to the data associated with the event that triggered the observer.
+    pub fn data_ptr(&self) -> Ptr {
+        Ptr::from(&self.data)
+    }
+
+    /// Returns the entity that triggered the observer, could be [`Entity::PLACEHOLDER`].
+    pub fn source(&self) -> Entity {
+        self.trigger.source
+    }
+}
+
+#[derive(Default, Clone)]
+pub(crate) struct ObserverDescriptor {
+    events: Vec<ComponentId>,
+    components: Vec<ComponentId>,
+    sources: Vec<Entity>,
+}
+
+/// Metadata for the source triggering an [`Observer`],
+pub struct ObserverTrigger {
+    observer: Entity,
+    event: ComponentId,
+    source: Entity,
+}
+
+// Map between an observer entity and it's runner
+type ObserverMap = EntityHashMap<Entity, ObserverRunner>;
+
+/// Collection of [`ObserverRunner`] for [`Observer`] registered to a particular event targeted at a specific component.
+#[derive(Default, Debug)]
+pub struct CachedComponentObservers {
+    // Observers listening to events targeting this component
+    map: ObserverMap,
+    // Observers listening to events targeting this component on a specific entity
+    entity_map: EntityHashMap<Entity, ObserverMap>,
+}
+
+/// Collection of [`ObserverRunner`] for [`Observer`] registered to a particular event.
+#[derive(Default, Debug)]
+pub struct CachedObservers {
+    // Observers listening for any time this event is fired
+    map: ObserverMap,
+    // Observers listening for this event fired at a specific component
+    component_observers: HashMap<ComponentId, CachedComponentObservers>,
+    // Observers listening for this event fired at a specific entity
+    entity_observers: EntityHashMap<Entity, ObserverMap>,
+}
+
+/// Metadata for observers. Stores a cache mapping event ids to the registered observers.
+#[derive(Default, Debug)]
+pub struct Observers {
+    // Cached ECS observers to save a lookup most common events.
+    on_add: CachedObservers,
+    on_insert: CachedObservers,
+    on_remove: CachedObservers,
+    // Map from event type to set of observers
+    cache: HashMap<ComponentId, CachedObservers>,
+}
+
+impl Observers {
+    pub(crate) fn get_observers(&mut self, event: ComponentId) -> &mut CachedObservers {
+        match event {
+            ON_ADD => &mut self.on_add,
+            ON_INSERT => &mut self.on_insert,
+            ON_REMOVE => &mut self.on_remove,
+            _ => self.cache.entry(event).or_default(),
+        }
+    }
+
+    pub(crate) fn try_get_observers(&self, event: ComponentId) -> Option<&CachedObservers> {
+        match event {
+            ON_ADD => Some(&self.on_add),
+            ON_INSERT => Some(&self.on_insert),
+            ON_REMOVE => Some(&self.on_remove),
+            _ => self.cache.get(&event),
+        }
+    }
+
+    pub(crate) fn invoke<E>(
```

Comment:
A small doc comment here explaining the purpose of this method would be useful: this seems to be the critical bit.